resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/CharsetManager.java

```diff
@@ -440,45 +440,60 @@ public void projectPreferencesChanged(IProject project) {
+		} else if (workspace.getRoot().findMember(resourcePath) instanceof Resource resource) {
+			setCharsetForResource(resourcePath, newCharset, resource);
 		}
-		// for all other cases, we set a property in the corresponding project
-		IResource resource = workspace.getRoot().findMember(resourcePath);
-		if (resource != null) {
-			try {
-				// disable the listener so we don't react to changes made by ourselves
-				Preferences encodingSettings = getPreferences(resource.getProject(), true, resource.isDerived(IResource.CHECK_ANCESTORS));
-				if (newCharset == null || newCharset.trim().length() == 0)
-					encodingSettings.remove(getKeyFor(resourcePath));
-				else
-					encodingSettings.put(getKeyFor(resourcePath), newCharset);
-				flushPreferences(encodingSettings, true);
-				if (resource instanceof IProject) {
-					IProject project = (IProject) resource;
-					ValidateProjectEncoding.scheduleProjectValidation(workspace, project);
-				}
-			} catch (BackingStoreException e) {
-				IProject project = workspace.getRoot().getProject(resourcePath.segment(0));
-				String message = Messages.resources_savingEncoding;
-				throw new ResourceException(IResourceStatus.FAILED_SETTING_CHARSET, project.getFullPath(), message, e);
+	}
+
+	private void setCharsetForRoot(IPath resourcePath, String newCharset) throws ResourceException {
+		IEclipsePreferences resourcesPreferences = InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES);
+		if (newCharset != null) {
+			resourcesPreferences.put(ResourcesPlugin.PREF_ENCODING, newCharset);
+		} else {
+			resourcesPreferences.remove(ResourcesPlugin.PREF_ENCODING);
+		}
+
+		try {
+			resourcesPreferences.flush();
+		} catch (BackingStoreException e) {
+			setCharsetForHasFailed(resourcePath, e);
+		}
+	}
+
+	private void setCharsetForResource(IPath resourcePath, String newCharset, IResource resource)
+			throws ResourceException {
+		try {
+			setResourceEncodingSettings(resourcePath, newCharset, resource);
+			if (resource instanceof Project project) {
+				ValidateProjectEncoding.scheduleProjectValidation(workspace, project);
 			}
+		} catch (BackingStoreException e) {
+			setCharsetForHasFailed(resourcePath, e);
 		}
 	}
 
+	private void setResourceEncodingSettings(IPath resourcePath, String newCharset, IResource resource)
+			throws BackingStoreException {
+		// disable the listener so we don't react to changes made by ourselves
+		Preferences encodingSettings = getPreferences(resource.getProject(), true,
+				resource.isDerived(IResource.CHECK_ANCESTORS));
+
+		if (newCharset == null || newCharset.isBlank()) {
+			encodingSettings.remove(getKeyFor(resourcePath));
+		} else {
+			encodingSettings.put(getKeyFor(resourcePath), newCharset);
+		}
+		flushPreferences(encodingSettings, true);
+	}
+
+	private void setCharsetForHasFailed(IPath resourcePath, BackingStoreException e) throws ResourceException {
+		IProject project = workspace.getRoot().getProject(resourcePath.segment(0));
```

Comment:
I resourcePath is root, the `segment(0)` will return `null`;  and `getProject` expects "The given name must be a valid path segment as defined by {@link IPath#isValidSegment(String)}." which is not the case. So this is likely to drive to some error.
I suggest the exception just contain the resourcePath, without trying to resolve it to a project.

Reply:
Your point is absolutely valid. Since this is pre-existing behavior, I would propose to have a separate PR fixing the behavior and not combine it with the pure refactoring of this PR (see https://github.com/eclipse-platform/eclipse.platform/pull/562#discussion_r1258181011).

Reply:
@mickaelistria The goal of this PR is not to change behavior but to refactor an existing function. I am fine with adding this behavior - I am not sure how having a separate commit is reconcileable with our aim to always have a single commit in each PR. Is such a case an exception to the rules?

Reply:
Having the fix as a separate commit without creating another PR is absolutely fine for me.

Reply:
There is no such requirementof having 1 commit per PR; the goal is to keep things efficient to develop and review. Depending on the change, that can mean different way of doing PRs/commits. Here I believe that an extra commit on this PR is the most efficient, but if you disagree, that's fine, it will be another PR later then.
Note that it's usually more critical to fix bugs than to refactor. Fixing a bug "by the way" is a good way to get refactorings merged in promptly ;)

Reply:
No, from my side there is no need to be dogmatic! I will add a separate commit :)